File: server/api/ollama.py
"""Ollama API endpoints for local AI management.

Provides endpoints for:
- Status: Check Ollama connection and model availability
- Models: List, pull, and manage models
- GPU: Detect and report GPU information
- Test: Test Ollama connection with custom URL
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from homebox_companion import settings
from homebox_companion.providers.ollama import OllamaProvider
from homebox_companion.services.gpu_detector import detect_gpu
from homebox_companion.services.ollama_manager import (
    OllamaManager,
    OllamaMode,
    get_ollama_manager,
)

router = APIRouter(prefix="/ollama")


# =============================================================================
# Request/Response Models
# =============================================================================


class OllamaTestRequest(BaseModel):
    """Request to test Ollama connection."""

    url: str | None = Field(None, description="Ollama URL to test (default: configured)")
    model: str | None = Field(None, description="Model to check for")


class OllamaTestResponse(BaseModel):
    """Response from Ollama connection test."""

    status: str
    connected: bool
    url: str
    model: str | None = None
    model_available: bool = False
    available_models: list[str] = []
    message: str | None = None


class ModelPullRequest(BaseModel):
    """Request to pull a model."""

    model: str = Field(..., description="Model name to pull")


class ModelPullResponse(BaseModel):
    """Response from model pull request."""

    status: str
    model: str
    error: str | None = None


class OllamaStatusResponse(BaseModel):
    """Full Ollama status response."""

    mode: str
    connected: bool
    url: str
    current_model: str
    model_ready: bool
    available_models: list[str]
    gpu: dict[str, Any] | None
    internal_running: bool
    error: str | None


class GPUInfoResponse(BaseModel):
    """GPU detection response."""

    detected: bool
    vendor: str
    name: str
    vram_mb: int
    vram_gb: float
    driver_version: str
    cuda_version: str
    compute_capability: str
    recommended_model: str
    supports_vision_models: bool
    detection_method: str


# =============================================================================
# Endpoints
# =============================================================================


@router.get("/status", response_model=OllamaStatusResponse)
async def get_ollama_status() -> OllamaStatusResponse:
    """Get current Ollama status.

    Returns connection status, model availability, and GPU information.
    """
    manager = await get_ollama_manager()
    status = await manager.get_status()
    return OllamaStatusResponse(**status.to_dict())


@router.post("/test", response_model=OllamaTestResponse)
async def test_ollama_connection(request: OllamaTestRequest) -> OllamaTestResponse:
    """Test connection to an Ollama server.

    Can test both the configured server and custom URLs.
    """
    url = request.url or settings.ollama_url
    model = request.model or settings.ollama_model

    provider = OllamaProvider(base_url=url, model=model)
    try:
        result = await provider.test_connection()
        return OllamaTestResponse(
            status=result.get("status", "unknown"),
            connected=result.get("connected", False),
            url=url,
            model=model,
            model_available=result.get("model_available", False),
            available_models=result.get("available_models", []),
            message=result.get("message"),
        )
    finally:
        await provider.close()


@router.get("/models", response_model=list[dict[str, Any]])
async def list_models() -> list[dict[str, Any]]:
    """List all available Ollama models."""
    manager = await get_ollama_manager()

    if manager.mode == OllamaMode.DISABLED:
        raise HTTPException(
            status_code=503,
            detail="Ollama is not enabled. Set HBC_USE_OLLAMA=true to enable.",
        )

    if not manager.provider:
        raise HTTPException(
            status_code=503,
            detail="Ollama is not initialized",
        )

    return await manager.list_models()


@router.post("/pull", response_model=ModelPullResponse)
async def pull_model(request: ModelPullRequest) -> ModelPullResponse:
    """Pull a model from Ollama registry.

    This may take several minutes for large models.
    """
    manager = await get_ollama_manager()

    if manager.mode == OllamaMode.DISABLED:
        raise HTTPException(
            status_code=503,
            detail="Ollama is not enabled. Set HBC_USE_OLLAMA=true to enable.",
        )

    result = await manager.pull_model(request.model)
    return ModelPullResponse(
        status=result.get("status", "unknown"),
        model=request.model,
        error=result.get("error"),
    )


@router.get("/gpu", response_model=GPUInfoResponse)
async def get_gpu_info() -> GPUInfoResponse:
    """Detect and return GPU information.

    Detects NVIDIA, AMD, Intel, and Apple Silicon GPUs.
    """
    gpu_info = detect_gpu()
    return GPUInfoResponse(**gpu_info.to_dict())


@router.get("/recommended-model")
async def get_recommended_model() -> dict[str, Any]:
    """Get recommended Ollama model based on detected GPU.

    Returns the model that best matches the available GPU VRAM.
    """
    gpu_info = detect_gpu()
    return {
        "recommended_model": gpu_info.recommended_model,
        "gpu_detected": gpu_info.detected,
        "gpu_name": gpu_info.name,
        "vram_gb": gpu_info.vram_gb,
        "supports_vision": gpu_info.supports_vision_models,
        "all_recommendations": {
            "48gb+": "llama3.2-vision:90b",
            "8gb+": "llama3.2-vision:11b",
            "6gb+": "minicpm-v",
            "4gb+": "granite3.2-vision",
            "3gb+": "moondream",
            "cpu": "moondream",
        },
    }


@router.get("/config")
async def get_ollama_config() -> dict[str, Any]:
    """Get current Ollama configuration from settings."""
    return {
        "use_ollama": settings.use_ollama,
        "ollama_internal": settings.ollama_internal,
        "ollama_url": settings.ollama_url,
        "ollama_model": settings.ollama_model,
        "fallback_to_cloud": settings.fallback_to_cloud,
    }

File: frontend/src/lib/api/ollama.ts
/**
 * Ollama API endpoints for local AI management
 */

import { request } from './client';

// =============================================================================
// TYPES
// =============================================================================

export interface OllamaStatus {
	mode: string;
	connected: boolean;
	url: string;
	current_model: string;
	model_ready: boolean;
	available_models: string[];
	gpu: GPUInfo | null;
	internal_running: boolean;
	error: string | null;
}

export interface OllamaTestRequest {
	url?: string;
	model?: string;
}

export interface OllamaTestResponse {
	status: string;
	connected: boolean;
	url: string;
	model: string | null;
	model_available: boolean;
	available_models: string[];
	message: string | null;
}

export interface GPUInfo {
	detected: boolean;
	vendor: string;
	name: string;
	vram_mb: number;
	vram_gb: number;
	driver_version: string;
	cuda_version: string;
	compute_capability: string;
	recommended_model: string;
	supports_vision_models: boolean;
	detection_method: string;
}

export interface ModelInfo {
	name: string;
	size: number;
	digest: string;
	modified_at: string;
	details?: {
		family: string;
		parameter_size: string;
		quantization_level: string;
	};
}

export interface ModelPullResponse {
	status: string;
	model: string;
	error: string | null;
}

export interface OllamaConfig {
	use_ollama: boolean;
	ollama_internal: boolean;
	ollama_url: string;
	ollama_model: string;
	fallback_to_cloud: boolean;
}

export interface RecommendedModelResponse {
	recommended_model: string;
	gpu_detected: boolean;
	gpu_name: string;
	vram_gb: number;
	supports_vision: boolean;
	all_recommendations: Record<string, string>;
}

// =============================================================================
// API FUNCTIONS
// =============================================================================

/**
 * Get current Ollama status including connection, model, and GPU info
 */
export const getOllamaStatus = () => request<OllamaStatus>('/ollama/status');

/**
 * Test connection to an Ollama server
 */
export const testOllamaConnection = (params?: OllamaTestRequest) =>
	request<OllamaTestResponse>('/ollama/test', {
		method: 'POST',
		body: JSON.stringify(params ?? {}),
	});

/**
 * List all available Ollama models
 */
export const listOllamaModels = () => request<ModelInfo[]>('/ollama/models');

/**
 * Pull a model from Ollama registry
 */
export const pullOllamaModel = (model: string) =>
	request<ModelPullResponse>('/ollama/pull', {
		method: 'POST',
		body: JSON.stringify({ model }),
	});

/**
 * Get GPU information
 */
export const getGPUInfo = () => request<GPUInfo>('/ollama/gpu');

/**
 * Get recommended model based on GPU
 */
export const getRecommendedModel = () =>
	request<RecommendedModelResponse>('/ollama/recommended-model');

/**
 * Get current Ollama configuration
 */
export const getOllamaConfig = () => request<OllamaConfig>('/ollama/config');
